File: Sources/GitBranchCleanerFfi/Result.swift
import Foundation

func encodeSuccess(_ value: Encodable) -> String {
  do {
    let data = try encode(from: value)
    return #"{"success": \#(data)}"#
  } catch {
    return ResultEncodingError(value).encoded()
  }
}

func encodeError(_ value: Error) -> String {
  do {
    let data =
      if case let value as Encodable = value {
        try encode(from: value)
      } else {
        String(describing: value)
      }
    return #"{"error": \#(data)}"#
  } catch {
    return ResultEncodingError(value).encoded()
  }
}

struct ResultEncodingError {
  let value: Any

  init(_ value: Any) {
    self.value = value
  }

  func encoded() -> String {
    let resultValue = String(describing: value)
    return #"{"error": {"type": "ResultEncodingError", "result": \#(resultValue)}}"#
  }
}

File: src/ffi/result.ts
export type Result<T> =
  | { type: "success"; value: T }
  | { type: "error"; value: object }
  | { type: "unknown"; value: any };

export function parseResult<T>(resultString: string): Result<T> {
  let resultData = JSON.parse(resultString);
  if ("success" in resultData) {
    return { type: "success", value: resultData["success"] as T };
  }
  if ("error" in resultData) {
    return { type: "error", value: resultData["error"] };
  }
  return { type: "unknown", value: resultData };
}
